🌱 Refactor and rename common part of transport controller

MikeSpreitzer opened the pull request:
<!--
Thanks for creating a pull request!

If this is your first time, please make sure to review CONTRIBUTING.MD.

Please copy the appropriate `:text:` or icon to the beginning of your PR title:

:sparkles: ✨ feature
:bug: 🐛 bug fix
:book: 📖 docs
:memo: 📝 proposal
:warning: ⚠️ breaking change
:seedling: 🌱 other/misc
:question: ❓ requires manual review/categorization

-->
## Summary
This PR moves the generic transport controller code that was in `cmd/` to `pkg/transport/cmd/`.

This PR also updates the architecture document to be a little more explicit about the structure here.

## Related issue(s)

Fixes #1750

MikeSpreitzer commented:
/cc @nirrozenbaum 
/cc @pdettori 
/cc @vMaroon

nirrozenbaum commented:
@MikeSpreitzer I'm ok with the idea in the PR but would like to request a few changes in the structure:
1. no need to keep the deprecated files under cmd. no one is using this (other than me) and I will fix the pointer in that ocm-plugin as soon as this PR will be merged. keeping deprecated is even more confusing than what it was before. 
let's just rename and move to pkg as you suggested. you don't break anything with this change cause nothing is running transport atm. we don't even break ocm plugin repo cause it's pointing to specific version of KS in go.mod (version before this PR).
2. the separation to `pkg/transport` and `pkg/transportcontroller` is strange. please let's put everything under `pkg/transport`. 
3. once everything is under `pkg/transport` just change the functions in `generic_transport_controller` to lowercase so it's not exposed, e.g., `func Run(...) -> func run(...)` and `func NewTransportController(...) -> func newTransportController(...)`. Then we can keep using function `Run(...)` in `base_controller`. calling it from outside as suggested in this PR looks like - `transportcontroller.MainForTransport()` - using transport twice is redundant, it's in the package name.
after this change it will be called `transport.Run(..)` or alternatively `transport.Main(..)`.   
I strongly prefer calling it `Run`, since as you stated, this is not executable under cmd and therefore it's not main and this also might be confusing.

MikeSpreitzer commented:
I deleted the deprecated stuff.

I changed NewTransportController to newTransportController.

I changed MainForTransport to GenericMain. There is still a problem with Run, in that a controller typically has a Run func --- and this one is no exception (genericTransportController::Run). Really we are using the word "controller" with two meanings here, one is the whole executable and one is the specific Go language object constructed by newTransportController. I think having the word "main" in the command func helps to direct the reader's thinking.

MikeSpreitzer commented:
I merged package transportcontroller into transport.

nirrozenbaum commented:
ok. there is a point in the the fact that a typical controller has Run function.
I can also live with GenericMain as the function name.
What if we put what was previously under `cmd/transport` into `pkg/transport/cmd`, would that work? 
it creates a clear separation between the "cmd-alike" part and the pkg classic part. wdyt?

MikeSpreitzer commented:
That package structure works for me too.

MikeSpreitzer commented:
The force-pushes make the latest package structure change and squash.

nirrozenbaum commented on docs/content/v0.20/architecture.md:
```diff
@@ -486,16 +486,16 @@ status updates for one object do not require updates of a whole bundle.
 
-In order to implement a KubeStellar transport plugin, one has to implement code that supplies a function to wrap any number of objects into a single wrapped objects. 
-The transport interface that should be implemented can be seen [here](https://github.com/kubestellar/kubestellar/blob/main/pkg/transport/transport.go).  
+Each plugin has an executable with a `main` func that calls [the generic code](../../pkg/transport/cmd/generic-main.go), passing the plugin object that implements the plugin interface.
```
broken link to the previous “transportcontroller”

nirrozenbaum commented:
it didn’t refresh so ignore the last comment.
/lgtm
Thanks!

MikeSpreitzer commented:
/approve
